File: sdk/android/core/src/main/kotlin/com/iapstack/core/IapStackConfig.kt
package com.iapstack.core

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import java.net.URI
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

/**
 * Runtime configuration for one authenticated mobile customer session.
 */
data class IapStackConfig(
  val baseUri: URI,
  val applicationId: String,
  val customerToken: String,
  val timeout: Duration = 10.seconds,
  val retryPolicy: IapStackRetryPolicy = IapStackRetryPolicy(),
  val maxResponseBytes: Int = 1024 * 1024,
  val allowInsecureHttp: Boolean = false,
) {
  /**
   * Parsed destination URL used by the transport layer.
   */
  val baseUrl: HttpUrl = baseUri.toString().toHttpUrl()

  init {
    validate()
  }

  /**
   * Throws when the configuration is unsafe or incomplete.
   */
  fun validate() {
    val host = baseUri.host ?: baseUri.authority
    if (host.isNullOrBlank() ||
      baseUri.query != null ||
      baseUri.fragment != null ||
      baseUri.userInfo.isNotEmpty()
    ) {
      throw IllegalArgumentException(
        "baseUri must be an origin or API path prefix without query, fragment, or user info",
      )
    }
    val scheme = baseUri.scheme.lowercase()
    if (scheme != "https" && !(allowInsecureHttp && scheme == "http")) {
      throw IllegalArgumentException("baseUri must use HTTPS")
    }
    if (applicationId.isBlank() || applicationId.contains('/')) {
      throw IllegalArgumentException("applicationId must be one non-empty path segment")
    }
    if (customerToken.isBlank() || customerToken.any { it.isWhitespace() }) {
      throw IllegalArgumentException("customerToken must be a non-empty bearer token")
    }
    if (timeout <= Duration.ZERO) {
      throw IllegalArgumentException("timeout must be positive")
    }
    if (maxResponseBytes <= 0) {
      throw IllegalArgumentException("maxResponseBytes must be positive")
    }
    retryPolicy.validate()
  }
}

File: sdk/ios/Sources/IAPStackApple/IAPStackConfig.swift
import Foundation

/// Runtime configuration for one authenticated customer session.
public struct IAPStackConfig: Sendable {
  /// Creates an immutable SDK configuration.
  public init(
    baseUri: URL,
    applicationId: String,
    customerToken: String,
    timeout: TimeInterval = 10,
    retryPolicy: IAPStackRetryPolicy = .init(),
    maxResponseBytes: Int = 1_048_576,
    allowInsecureHttp: Bool = false,
  ) {
    self.baseUri = baseUri
    self.applicationId = applicationId
    self.customerToken = customerToken
    self.timeout = timeout
    self.retryPolicy = retryPolicy
    self.maxResponseBytes = maxResponseBytes
    self.allowInsecureHttp = allowInsecureHttp
  }

  /// IAPStack origin, optionally including a path prefix.
  public let baseUri: URL

  /// Application scope encoded in every public API path.
  public let applicationId: String

  /// Short-lived customer token retained only in-memory.
  public let customerToken: String

  /// Max duration for one network attempt.
  public let timeout: TimeInterval

  /// Bounded retry policy for idempotent operations.
  public let retryPolicy: IAPStackRetryPolicy

  /// Maximum accepted response size.
  public let maxResponseBytes: Int

  /// Allows plain HTTP for local development only.
  public let allowInsecureHttp: Bool

  /// Throws an `IAPStackSDKError.configurationError` on unsafe configuration.
  public func validate() throws {
    guard let host = baseUri.host, !host.isEmpty else {
      throw IAPStackSDKError.configurationError(message: "baseUri must include a valid host")
    }
    if baseUri.query != nil || baseUri.fragment != nil || baseUri.user != nil || baseUri.password != nil {
      throw IAPStackSDKError.configurationError(
        message: "baseUri must be an origin or path prefix without query, fragment, or credentials",
      )
    }
    let scheme = baseUri.scheme?.lowercased() ?? ""
    if scheme != "https" && !(allowInsecureHttp && scheme == "http") {
      throw IAPStackSDKError.configurationError(message: "baseUri must use HTTPS")
    }
    if applicationId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
      applicationId.contains("/") {
      throw IAPStackSDKError.configurationError(
        message: "applicationId must be one non-empty path segment",
      )
    }
    if customerToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
      customerToken.contains(where: { $0.isWhitespace }) {
      throw IAPStackSDKError.configurationError(
        message: "customerToken must be a non-empty bearer token",
      )
    }
    if timeout <= 0 {
      throw IAPStackSDKError.configurationError(message: "timeout must be positive")
    }
    if maxResponseBytes <= 0 {
      throw IAPStackSDKError.configurationError(
        message: "maxResponseBytes must be positive",
      )
    }
    try retryPolicy.validate()
  }
}
